### ISSUE TITLE
When I created a flinkcluster, add a watch to the cluster .and get the watch data with 'CountDownLatch.wait()''

### ISSUE THREAD

**Issue by Mrart** [GoogleCloudPlatform/flink-on-k8s-operator#187]
2020-03-02T12:34:56.652Z	DEBUG	controller-runtime.manager.events	Normal	{"object": {"kind":"FlinkCluster","namespace":"sloth","name":"sloth-818","uid":"2cb88c0c-5c82-11ea-b628-246e9613cf8c","apiVersion":"flinkoperator.k8s.io/v1beta1","resourceVersion":"6340425"}, "reason": "StatusUpdate", "message": "JobManager deployment status changed: Deleted -> NotReady"}
2020-03-02T12:34:56.655Z	ERROR	controllers.FlinkCluster	Failed to update cluster status	{"cluster": "sloth/sloth-818", "error": "Operation cannot be fulfilled on flinkclusters.flinkoperator.k8s.io \"sloth-818\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
	/root/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/googlecloudplatform/flink-operator/controllers.(*FlinkClusterHandler).reconcile
	/workspace/controllers/flinkcluster_controller.go:151
github.com/googlecloudplatform/flink-operator/controllers.(*FlinkClusterReconciler).Reconcile
	/workspace/controllers/flinkcluster_controller.go:76
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:216
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:192
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:171
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/root/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/root/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153
k8s.io/apimachinery/pkg/util/wait.Until
	/root/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88
2020-03-02T12:34:56.655Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "flinkcluster", "request": "sloth/sloth-818", "error": "Operation cannot be fulfilled on flinkclusters.flinkoperator.k8s.io \"sloth-818\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
	/root/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:218
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:192
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/root/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.2/pkg/internal/controller/controller.go:171
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/root/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/root/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153
k8s.io/apimachinery/pkg/util/wait.Until
	/root/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88

**Comment by functicons** [comment]
The error `the object has been modified; please apply your changes to the latest version and try again` is transient, usually it should recover automatically.

**Comment by Mrart** [comment]
@functicons  It seem can recover, but the java client has recieved a '409' like `Create flink clusters error...Failure executing: POST at: https://bigdata-test-k8s-api.netease.com:6443/apis/flinkoperator.k8s.io/v1beta1/namespaces/sloth/flinkclusters. Message: flinkclusters.flinkoperator.k8s.io "sloth-818" already exists. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=flinkoperator.k8s.io, kind=flinkclusters, name=sloth-818, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=flinkclusters.flinkoperator.k8s.io "sloth-818" already exists, metadata=ListMeta(_continue=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=AlreadyExists, status=Failure, additionalProperties={}).
	at com.netease.sloth.kernel.SlothK8SOperatorService.startJobAsync(SlothK8SOperatorService.java:86)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.netease.flick.server.DefaultServiceContainer$Procedure.call(DefaultServiceContainer.java:139)
	at com.netease.flick.server.DefaultServiceContainer$RpcTask.run(DefaultServiceContainer.java:191)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)`

**Comment by functicons** [comment]
Are you trying to create a flink job cluster with Java API? The error message says `flinkclusters.flinkoperator.k8s.io "sloth-818" already exists`, it is not the same problem as `the object has been modified; please apply your changes to the latest version and try again`.

**Comment by Mrart** [comment]
Ok, It other bug.

**Comment by Mrart** [comment]
@functicons  I get this error again ,can you explain why got this error log.

**Comment by Mrart** [comment]
I reopen it, because I think i caused some error , about the status.

**Comment by functicons** [comment]
The error `as the object has been modified; please apply your changes to the latest version and try again.` happens when we are trying to update the status of CR concurrently, considering this sequence:

1) controller sends status update 1 to API server asynchronously
2) controller sends status update 2 to API server asynchronously
3) status update 1 is done on the server side
4) update 2 sees the error "the object has been modified; please apply your changes to the latest version and try again".

basically, this error should simply be retried in the next reconciliation iteration.
